### PR TITLE
[SYCL][Graph] Enable specialization constants with graph

### PR DESCRIPTION
**Pull request by EwanC**
- Enables specialization constants handling in SYCL-Graph extension.
- Adds E2E tests that verify this behavior.
- Removes unittests tests that checked for unsupported feature exception throwing.

### CODE REVIEW

**Comment by bader** [comment]
"specification constants" -> "specialization constants"

**Comment by EwanC** [comment]
> "specification constants" -> "specialization constants"

Thanks for spotting 🤦 Updated PR title and description,

File: sycl/source/handler.cpp [hunk 1]
```diff
@@ -179,10 +180,10 @@ event handler::finalize() {
       // Make sure implicit non-interop kernel bundles have the kernel
       if (!KernelBundleImpPtr->isInterop() &&
           !MImpl->isStateExplicitKernelBundle()) {
+        auto Dev = (MGraph) ? MGraph->getDevice() : MQueue->get_device();
```
Comment (Bensuo) [inline sycl/source/handler.cpp:183]:
```suggestion
        auto Dev = MGraph ? MGraph->getDevice() : MQueue->get_device();
```
Nit but the parenthesis are unnecessary here.

**Review by Bensuo — APPROVED** [review]

File: sycl/source/handler.cpp [hunk 2]
```diff
@@ -111,9 +111,10 @@ bool handler::isStateExplicitKernelBundle() const {
 std::shared_ptr<detail::kernel_bundle_impl>
 handler::getOrInsertHandlerKernelBundle(bool Insert) const {
   if (!MImpl->MKernelBundle && Insert) {
-    MImpl->MKernelBundle =
-        detail::getSyclObjImpl(get_kernel_bundle<bundle_state::input>(
-            MQueue->get_context(), {MQueue->get_device()}, {}));
+    auto Ctx = (MGraph) ? MGraph->getContext() : MQueue->get_context();
+    auto Dev = (MGraph) ? MGraph->getDevice() : MQueue->get_device();
```
Comment (reble) [inline sycl/source/handler.cpp:115]: Redundant brackets here too, like in Ben's comment below:
https://github.com/intel/llvm/pull/11556/files#r1365333849

**Comment by EwanC** [comment]
@intel/llvm-reviewers-runtime Could this have a review please

**Comment by EwanC** [comment]
ping @intel/llvm-reviewers-runtime

**Comment by EwanC** [comment]
@intel/llvm-reviewers-runtime Can you review this please

File: sycl/source/handler.cpp [hunk 3]
```diff
@@ -111,9 +111,10 @@ bool handler::isStateExplicitKernelBundle() const {
 std::shared_ptr<detail::kernel_bundle_impl>
 handler::getOrInsertHandlerKernelBundle(bool Insert) const {
   if (!MImpl->MKernelBundle && Insert) {
-    MImpl->MKernelBundle =
-        detail::getSyclObjImpl(get_kernel_bundle<bundle_state::input>(
-            MQueue->get_context(), {MQueue->get_device()}, {}));
+    auto Ctx = MGraph ? MGraph->getContext() : MQueue->get_context();
+    auto Dev = MGraph ? MGraph->getDevice() : MQueue->get_device();
```
Comment (steffenlarsen) [inline sycl/source/handler.cpp:115]:
```suggestion
    auto &Ctx = MGraph ? MGraph->getContext() : MQueue->get_context();
    auto &Dev = MGraph ? MGraph->getDevice() : MQueue->get_device();
```
If both return references? Same applies other places if applicable here.

File: sycl/test-e2e/Graph/Inputs/spec_constants_handler_api.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    Queue.submit([&](handler &CGH) { CGH.ext_oneapi_graph(GraphExec); });
+    Queue.wait_and_throw();
+  }
+
+  sycl::host_accessor IntAcc(IntBuffer, sycl::read_only);
+  if (!check_value(
+          0, IntAcc[0],
+          "integer specialization constant (defined without default value)"))
+    return false;
```
Comment (steffenlarsen) [inline sycl/test-e2e/Graph/Inputs/spec_constants_handler_api.cpp:98]: Might it be useful to let the other test cases run and fail too, just to have as much information as possible or are they too closely tied that it won't give much value?

File: sycl/source/handler.cpp (hunk 3 above)
Comment (EwanC) [reply sycl/source/handler.cpp:115]: None of these actually return references, the first returns a `sycl::context` and the second `sycl::device`, both by value.

File: sycl/test-e2e/Graph/Inputs/spec_constants_handler_api.cpp (hunk 4 above)
Comment (EwanC) [reply sycl/test-e2e/Graph/Inputs/spec_constants_handler_api.cpp]: Good idea, updated to try the following tests rather than exiting early.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by EwanC** [comment]
@intel/llvm-gatekeepers I think this PR is good to merge, the CI is green and has approvals from necessary groups

**Pull request merged**